src/builder.ts

```diff
@@ -95,56 +106,58 @@ export default class Builder {
 							if (data == null) {
 								return;
 							}
-							if (data.error) {
-								errored = true;
-								dup.destroy(new Error(data.error));
-							} else {
-								// Store image layers, so that they can be deleted by the caller
-								// if necessary
-								const sha = Utils.extractLayer(data.stream);
-								if (sha !== undefined) {
-									self.layers.push(sha);
+							try {
+								if (data.error) {
+									throw new Error(data.error);
+								} else {
+									// Store image layers, so that they can be
+									// deleted by the caller if necessary
+									const sha = Utils.extractLayer(data.stream);
+									if (sha !== undefined) {
+										self.layers.push(sha);
+									}
+									this.emit('data', data.stream);
 								}
-
-								this.emit('data', data.stream);
+							} catch (error) {
+								daemonStream.unpipe();
+								reject(error);
 							}
 						}),
 					);
 
-				// Catch any errors the stream produces
-				outputStream.on('error', (err: Error) => {
-					errored = true;
-					self.callHook(hooks, 'buildFailure', handler, err, self.layers);
-				});
-				dup.on('error', (err: Error) => {
-					errored = true;
-					self.callHook(hooks, 'buildFailure', handler, err, self.layers);
-				});
-
-				// Setup the buildSuccess hook. This handler is not called on
-				// error so we can use it to propagate the success information
-				outputStream.on('end', () => {
-					if (!errored) {
-						this.callHook(
-							hooks,
-							'buildSuccess',
-							handler,
-							_.last(this.layers),
-							this.layers,
-						);
-					}
+				outputStream.on('error', (error: Error) => {
+					daemonStream.unpipe();
+					reject(error);
 				});
+				outputStream.on('end', () =>
+					// The 'end' event was observed to be emitted under error
+					// conditions, hence the test for streamError.
+					streamError ? reject(streamError) : resolve(),
+				);
 				// Connect the output of the docker daemon to the duplex stream
 				dup.setReadable(outputStream);
-			})
-			.catch((err: Error) => {
-				// Call the plugin's error handler
-				self.callHook(hooks, 'buildFailure', handler, err, self.layers);
 			});
+		});
+
+		Bluebird.join(
+			buildPromise,
+			// Call the buildStream handler with the docker daemon stream
+			this.callHook(hooks, 'buildStream', handler, dup),
+		)
+			.then(() => {
+				if (!streamError) {
+					// Build successful: call buildSuccess handler
+					return this.callHook(
+						hooks,
+						'buildSuccess',
+						handler,
+						_.last(this.layers),
+						this.layers,
+					);
+				}
+			})
+			.catch(failBuild);
 
-		// Call the correct hook with the build stream
-		this.callHook(hooks, 'buildStream', handler, dup);
```

Comment:
The promise returned by this `callHook` call was being ignored in case it was rejected (`handler` would be called with an error, but it is an optional argument). I have moved this hook call to a `Bluebird.join`, so the build fails if the promise returned by the hook call is rejected.

Reply:
I don't see how `join` could help here.

```typescript
buildPromise
    .then(this.callHook(hooks, 'buildStream', handler, dup))
    .then(() => {
        if (!streamError) {
            return this.callHook(hooks, 'buildSuccess', handler, _.last(this.layers), this.layers)
        }
    })
    .catch(failBuild);
```

or more readable:

```typescript
try {
    await buildPromise;
    await this.callHook(hooks, 'buildStream', handler, dup);
    if (!streamError) {
        await this.callHook(hooks, 'buildSuccess', handler, _.last(this.layers), this.layers)
    }
} catch (error) {
    failBuild(error);
}
```

Also, `callHook` should probably reject when it can't find the hook instead of calling handler and resolving.

Reply:
There is a good reason, actually! The buildStream hook must be called in order for buildPromise to ever resolve, so at least it would have to be awaited first (before buildPromise). This is because the buildStream hook consumes the stream whose 'end' even resolves buildPromise. But buildPromise may also reject before the stream is fully consumed (in which case the stream in indeed unpipe'd and destroy'ed), so it is beneficial to await for both promises in parallel and reject as soon as either rejects, which is what `Bluebird.join()` achieves.

I have now added a comment to the code, to make this clearer. I have also realised that `createBuildStream` was getting large, so I've now refactored a chunk of it to a separate function (`getDockerDaemonBuildOutputParserStream`).

Reply:
You're calling `Bluebird.join` with 2 parameters.
http://bluebirdjs.com/docs/api/promise.join.html
>The final parameter, handler function, will be invoked with the result values of all of the fulfilled promises

There's a note on the end of the page on why this still works (not passing a function as last argument) but it is deprecated.
>Note: In 1.x and 0.x Promise.join used to be a Promise.all that took the values in as arguments instead of an array. This behavior has been deprecated but is still supported partially - when the last argument is an immediate function value the new semantics will apply

Reply:
I understood the code wrong in my first comments because of `join` parameters.
```typescript
try {
    await [ buildPromise, this.callHook(hooks, 'buildStream', handler, dup)];
    if (!streamError) {
        await this.callHook(hooks, 'buildSuccess', handler, _.last(this.layers), this.layers)
    }
} catch (error) {
    failBuild(error);
}
```

Reply:
Thanks, @zvin! I have replaced `Bluebird.join` with `Bluebird.all`. `createBuildStream` is a synchronous function that returns a stream, so using `await` would require some sort of wrapping of the code. By the way, `await` with an array, does that really work? I haven't tested it, but MDN and StackOverflow suggest it does not:
* _"If the value of the expression following the await operator is not a Promise, it's converted to a resolved Promise."_ https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await
* https://stackoverflow.com/questions/46889290/waiting-for-more-than-one-concurrent-await-operation

As if it was the same as `await Promise.resolve([ promise1, promise2])`, which I guess would not really await for each promise in the array.

Reply:
Sorry I meant `await Promise.all([ promise1, promise2 ])`.
```typescript
try {
    await Promise.all([ buildPromise, this.callHook(hooks, 'buildStream', handler, dup) ]);
    if (!streamError) {
        await this.callHook(hooks, 'buildSuccess', handler, _.last(this.layers), this.layers)
    }
} catch (error) {
    failBuild(error);
}
```

Reply:
If you want to run this in parallel:
```typescript
(async () => {
    try {
        await Promise.all([ buildPromise, this.callHook(hooks, 'buildStream', handler, dup) ]);
        if (!streamError) {
            await this.callHook(hooks, 'buildSuccess', handler, _.last(this.layers), this.layers)
        }
    } catch (error) {
        failBuild(error);
    }
})()
```

Reply:
Yeah, that `(async () => { ... })();` wrapping, plus the use of `Promise.all` inside, kind of spoils the simplicity of `await`. I'll stick with `Bluebird.all().then();` for this one. :-)